src/kernel-install/60-ukify.install.in

```diff
@@ -0,0 +1,203 @@
+
+    log('KERNEL_INSTALL_LAYOUT and KERNEL_INSTALL_UKI_GENERATOR are good')
+    return True
+
+def kernel_cmdline() -> str:
+    entry_token = os.environ['KERNEL_INSTALL_ENTRY_TOKEN']
+    machine_id = os.environ['KERNEL_INSTALL_MACHINE_ID']
+
+    if root := os.getenv('KERNEL_INSTALL_CONF_ROOT'):
+        try:
+            options = Path(root).joinpath('cmdline').read_text().split()
+        except FileNotFoundError:
+            return None
+    else:
+        for cmdline, fallback in [
+                ('/etc/kernel/cmdline',     False),
+                ('/usr/lib/kernel/cmdline', False),
+                ('/proc/cmdline',           True),
+        ]:
+            try:
+                options = Path(cmdline).read_text().split()
+            except FileNotFoundError:
+                if fallback:
+                    raise
+
+        if fallback:
+            options = [opt for opt in options
+                       if not opt.startswith(('BOOT_IMAGE=', 'initrd='))]
+
+    # If the boot entries are named after the machine ID, then suffix the kernel
+    # command line with the machine ID we use, so that the machine ID remains
+    # stable, even during factory reset, in the initrd (where the system's machine
+    # ID is not directly accessible yet), and if the root file system is volatile.
+    if (entry_token == machine_id and
+        not any(opt.startswith('systemd.machine_id=') for opt in options)):
+        options += [f'systemd.machine_id={machine_id}']
+
+    cmdline = ' '.join(options)
+    if cmdline.startswith('@'):
+        raise ValueError("Kernel options must not start with '@'")
+    return cmdline
+
+def make_uki(opts, cmdline):
+    output = opts.staging_area / 'uki.efi'
+
+    # Args must be strings here to not confuse argparse in ukify.parse_args().
+    args = [
+        opts.kernel_image,
+        *opts.initrd,
+        f'--uname={opts.kernel_version}',
+        f'--output={output}',
+        f'--cmdline={cmdline}',
+    ]
+
+    if BOOT_STUB:
+        args += [f'--stub={BOOT_STUB}']
+
+    # Punish me harder.
+    # We want this:
+    #   ukify = importlib.machinery.SourceFileLoader('ukify', UKIFY).load_module()
+    # but it throws a DeprecationWarning.
+    # https://stackoverflow.com/questions/67631/how-can-i-import-a-module-dynamically-given-the-full-path
+    # https://github.com/python/cpython/issues/65635
+    # offer "explanations", but to actually load a python file without a .py extension,
+    # the "solution" is 4+ incomprehensible lines.
+    # The solution with runpy gives a dictionary, which isn't great, but will do.
+    ukify = runpy.run_path(UKIFY, run_name='ukify')
+    ukify['main'](args)
+
+    log(f'{output} has been created')
+
+def main():
+    opts = parse_args()
+    if opts.command != 'add':
+        return
+    if not we_are_wanted():
+        return
+
+    cmdline = kernel_cmdline()
+    if cmdline is None:
+        sys.exit('No cmdline configuration found')
+
+    make_uki(opts, cmdline)
```

Comment:
Right now raised errors are never caught and will thus produce tracebacks if something goes wrong, the exit code will also ever be 1. 

Adding a custom error and wrapping the body of `main` in a try block, could be used to suppress that
```suggestion
class UkifyInstallError(Exception):
    def __init__(self, msg, rc=1):
        self.msg = msg
        self.rc = rc
    

def main():
    try:
        opts = parse_args()
        if opts.command != 'add':
            return
        if not we_are_wanted():
            return

        cmdline = kernel_cmdline()
        if cmdline is None:
            sys.exit('No cmdline configuration found')

        make_uki(opts, cmdline)
    except UkifyInstallError as e:
        print(e.msg, file=sys.stderr)
        sys.exit(e.rc)
```
This way some of the errors could then also get an exit code of 77 and `kernel_cmdline()` could throw the error instead of returning `None`. We would also see whether an exit was from an error we already anticipated or was a bug. This would also leave the door open to add some debug environment variable to always print the stack trace.

When going this route, the raises should look like, e.g.
```python
except ValueError as e:
    raise UkifyInstallError("Kernel options must not start with '@'", 77) from e
```

Reply:
Yeah, I think it makes sense to suppress tracebacks. But I'm leaving that for later… As the discussion in mkosi shows, getting this just right is quite hard, and generally during early development its convenient to rather have too many tracebacks than too few.